cirq/sim/simulator.py

```diff
@@ -447,9 +448,134 @@ def _simulator_iterator(
+                define the order of amplitudes in the wave function.
+            initial_state: If an int, the state is set to the computational
+                basis state corresponding to this state.
+                Otherwise if this is a np.ndarray it is the full initial state.
+                In this case it must be the correct size, be normalized (an L2
+                norm of 1), and  be safely castable to an appropriate
+                dtype for the simulator.
+
+        Returns:
+            List of ComputeDisplaysResults for this run, one for each
+            possible parameter resolver.
+        """
+        circuit = (program if isinstance(program, circuits.Circuit)
+                   else program.to_circuit())
+        param_resolvers = study.to_resolvers(params or study.ParamResolver({}))
+        qubit_order = ops.QubitOrder.as_qubit_order(qubit_order)
+        qubits = qubit_order.order_for(circuit.all_qubits())
+
+        compute_displays_results = []  # type: List[study.ComputeDisplaysResult]
+        for param_resolver in param_resolvers:
+            display_values = {}  # type: ignore
+
+            # Compute the displays in the first Moment
+            moment = circuit[0]
+            state = wave_function.to_valid_state_vector(
+                initial_state, num_qubits=len(qubits))
+            qubit_map = {q: i for i, q in enumerate(qubits)}
+            _enter_moment_display_values_into_dictionary(
+                display_values, moment, state, qubit_order, qubit_map)
+
+            # Compute the displays in the rest of the Moments
+            all_step_results = self.simulate_moment_steps(circuit,
+                                                          param_resolver,
+                                                          qubit_order,
+                                                          initial_state)
+            for step_result, moment in zip(all_step_results, circuit[1:]):
+                _enter_moment_display_values_into_dictionary(
+                    display_values,
+                    moment,
+                    step_result.state(),
+                    qubit_order,
+                    step_result.qubit_map)
+
+            compute_displays_results.append(study.ComputeDisplaysResult(
+                params=param_resolver,
+                display_values=display_values))
+
+        return compute_displays_results
+
+
+def _enter_moment_display_values_into_dictionary(
+        display_values: Dict,
+        moment: ops.Moment,
+        state: np.ndarray,
+        qubit_order: ops.QubitOrder,
+        qubit_map: Dict[ops.QubitId, int]):
+    displays = (op for op in moment
+                if isinstance(op, (ops.SamplesDisplay,
+                                   ops.WaveFunctionDisplay)))
+    for display in displays:
+        if isinstance(display, ops.WaveFunctionDisplay):
+            display_values[display.key] = (
```

Comment:
Does this correctly account for the fact that, if there are measurements before this display, it is necessary to sample?

We might want a 'density matrix display' concept for the more general case, at some point.

Reply:
I don't understand. Measurements before this display can only be in a previous moment. Why would it be necessary to sample?

Reply:
Ah, I see, you mean that the value of some displays (like idealized pauli string expectations) should not depend on the outcome of previous measurements, and rather take the value it would have from a density matrix simulations. Hmm ...

Reply:
What do you think of adding a `repetitions` argument to `compute_displays`, indicating to average the value of displays over that many circuit repetitions?

Reply:
I think that should be a per-display option, not a global option.